File: 3.longest-substring-without-repeating-characters.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int le = s.length();
        int res = 0;
        Map<Character, Integer> map = new HashMap<>();
        for (int j = 0, i = 0; j < le; j++) {
            if (map.containsKey(s.charAt(j))) {
                i = Math.max(map.get(s.charAt(j)), i);
            }
            map.put(s.charAt(j), j + 1);
            res = Math.max(j - i + 1, res);
        }
        return res;
    }
}

File: 3.longest-substring-without-repeating-characters.cs
/*
 * @lc app=leetcode id=3 lang=csharp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
public class Solution
{
    public int LengthOfLongestSubstring(string s) {
        int left = 0, right = 0;
        int res = 0;
        var dic = new Dictionary<char, int>();
        for(right = 1; right <= s.Length; right++) {
            var curChar = s[right - 1];
            if(dic.ContainsKey(curChar)) {
                if(dic[curChar] >= left)
                    left = dic[curChar] + 1;
                dic[curChar] = right - 1;
            }
            else {
                dic.Add(curChar, right - 1);
            }
            if(res < right - left)
                res = right - left;
        }
        return res;
    }
}
// @lc code=end

